timemachine/md/hrex.py

```diff
@@ -0,0 +1,247 @@
+from dataclasses import dataclass
+from typing import Callable, Generic, List, NewType, Optional, Sequence, Tuple, TypeVar
+
+import numpy as np
+from numpy.typing import NDArray
+
+from timemachine.md.moves import MetropolisHastingsMove, MixtureOfMoves
+from timemachine.utils import batches
+
+_Replica = TypeVar("_Replica")
+
+StateIdx = NewType("StateIdx", int)
+ReplicaIdx = NewType("ReplicaIdx", int)
+
+
+@dataclass
+class NeighborSwapMove(MetropolisHastingsMove[List[_Replica]]):
+    def __init__(self, log_q: Callable[[_Replica, StateIdx], float], s_a: StateIdx, s_b: StateIdx):
+        super().__init__()
+        self.log_q = log_q
+        self.s_a = s_a
+        self.s_b = s_b
+
+    def propose_with_log_q_diff(self, state: List[_Replica]) -> Tuple[List[_Replica], float]:
+        s_a = self.s_a
+        s_b = self.s_b
+        state_ = list(state)
+        state_[s_a], state_[s_b] = state[s_b], state[s_a]
+        proposed_state = state_
+
+        r_a = state[s_a]
+        r_b = state[s_b]
+        log_q_diff = self.log_q(r_a, s_b) + self.log_q(r_b, s_a) - self.log_q(r_a, s_a) - self.log_q(r_b, s_b)
+
+        return proposed_state, log_q_diff
+
+
+_Samples = TypeVar("_Samples")
+
+
+@dataclass
+class Hrex(Generic[_Replica]):
```

Comment:
we use HREX (in full upper case) a lot, maybe we should just use all-caps here?

Reply:
Could do; I'd chosen this convention because multi-word identifiers can look awkward with the all-caps style (e.g. `HREXDiagnostics`), but have no strong preference.

Reply:
Renamed in https://github.com/proteneer/timemachine/pull/1128/commits/2878505cfe1d8b0de5cdffc430203def182a72e8